### PR TITLE
remove schedule global lock

### PR DESCRIPTION
**Pull request by ShellyWEI**
Previously Schedule system's lock is pretty strict.
- synchronized keywords ensures only one user would getSchedule, and even if viewing a single schedule would require updates the whole schedules.
- syncObj make sure only one trigger can be updated at a time, and it's conflicting with trigger scanning process. 

These locks makes schedule system unable to handle large load as only one of hundreds would acquire the lock and do the operation while all other requests would need to wait for the lock to release, even if those requests are totally irrelevant with each other. (One should be able to update its schedule, while less rely on other schedule updates) 

This PR removes all sync keywords and the global lock to lower the lock level to individual schedule/triggers instead of keeping all of them in the global synchronized stage. ScheduleManager is the entry point that links schedule and trigger, it will maintain its own checkTime for each individual trigger with its lastModifyTime as evidence to check if a schedule/trigger is getting updated.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/scheduler/ScheduleManager.java
Comment (tsangz2013) [inline azkaban-common/src/main/java/azkaban/scheduler/ScheduleManager.java:272]: Not super familiar with Java `finally`, I think move `return s;` to the end of method is more easy to read

File: azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java
Comment (tsangz2013) [inline azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java:325]: Can you explain a little about the old dynamic of `this.syncObj` and why it's safe to directly sleep the current Thread instead?

File: azkaban-common/src/main/java/azkaban/scheduler/Schedule.java
Comment (ShellyWEI) [reply azkaban-common/src/main/java/azkaban/scheduler/Schedule.java:252]: check the ReentrantLock's java doc. This method does not throw any exceptions. concurrent package usually operated atomically on JVM directly.

File: azkaban-common/src/main/java/azkaban/scheduler/ScheduleManager.java
Comment (ShellyWEI) [reply azkaban-common/src/main/java/azkaban/scheduler/ScheduleManager.java:272]: If anything happens before "returnn s", for example an exception happens. this method might exit but this lock would never release. So in java, it's recommended to use "finally" keywords to release lock.

File: azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java
Comment (ShellyWEI) [reply azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java:325]: Again it's a good habit to check java doc directly. An obj's wait method is a direct waiting to be invoked operation, anything interruption would awake this obj to continue with the code. -> of course in the original implementation, there are keywords "synchrinozed", which means others need to wait when someone ontain this obj's lock. But I remove them in this PR, when this syncObj only remains the "sleep" functionality.

File: azkaban-common/src/main/java/azkaban/scheduler/ScheduleManager.java
Comment (tsangz2013) [reply azkaban-common/src/main/java/azkaban/scheduler/ScheduleManager.java:272]: in current setting, if no exception, it will unlock and then return? or it will return but also unlock?

**Review by tsangz2013 — APPROVED** [review]
lgtm, please test carefully before merging

File: azkaban-common/src/main/java/azkaban/scheduler/ScheduleManager.java
Comment (ShellyWEI) [reply azkaban-common/src/main/java/azkaban/scheduler/ScheduleManager.java:272]: yes, using "finally" means after the "try" it will whatever unblock. If it successfully executes return it will return.

**Pull request merged**
